extensions/amp-orientation-observer/0.1/amp-orientation-observer.js

```diff
@@ -0,0 +1,175 @@
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private @const */
+    this.ampdoc_ = Services.ampdoc(this.element);
+
+    /** @private {?../../../src/service/action-impl.ActionService} */
+    this.action_ = null;
+
+    /** @private {Array<number>} */
+    this.alphaRange_ = [0, 360];
+
+    /** @private {Array<number>} */
+    this.betaRange_ = [-180, 180];
+
+    /** @private {Array<number>} */
+    this.gammaRange_ = [-90, 90];
+
+    /** @private {number} */
+    this.alphaValue_ = 180; // Set to the neutral resting position
+
+    /** @private {number} */
+    this.betaValue_ = 0; // Set to the neutral resting position
+
+    /** @private {number} */
+    this.gammaValue_ = 0; // Set to the neutral resting position
+  }
+
+  /** @override */
+  buildCallback() {
+    // Since this is a functional component and not visual,
+    // layoutCallback is meaningless. We delay the heavy work until
+    // we become visible.
+    this.action_ = Services.actionServiceForDoc(this.element);
+    const viewer = Services.viewerForDoc(this.ampdoc_);
+    viewer.whenFirstVisible().then(this.init_.bind(this));
+  }
+
+  /**
+   * @private
+   */
+  init_() {
+    this.parseAttributes_();
+
+    user().assert(this.win.DeviceOrientationEvent,
+        'The current browser doesn\'t support the ' +
+      '`window.DeviceOrientationEvent`');
+    this.win.addEventListener('deviceorientation', event => {
+      this.deviceOrientationHandler_(event);
+    }, true);
+  }
+
+  /**
+   * @private
+   */
+  parseAttributes_() {
+    const alphaRange = this.element.getAttribute('alpha-range');
+    const betaRange = this.element.getAttribute('beta-range');
+    const gammaRange = this.element.getAttribute('gamma-range');
+    if (alphaRange) {
+      const range = alphaRange.trim().split(' ');
+      this.alphaRange_ = [parseInt(range[0], 10), parseInt(range[1], 10)];
+    }
+    if (betaRange) {
+      const range = betaRange.trim().split(' ');
+      this.betaRange_ = [parseInt(range[0], 10), parseInt(range[1], 10)];
+    }
+    if (gammaRange) {
+      const range = gammaRange.trim().split(' ');
+      this.gammaRange_ = [parseInt(range[0], 10), parseInt(range[1], 10)];
+    }
+  }
+
+  /**
+   * @param {!Event} event
+   * @private
+   */
+  deviceOrientationHandler_(event) {
+    if (event instanceof DeviceOrientationEvent) {
+      if (event.alpha !== this.alphaValue_) {
+        this.alphaValue_ = event.alpha;
+        this.triggerAlpha_();
+      }
+      if (event.beta !== this.betaValue_) {
+        this.betaValue_ = event.beta;
+        this.triggerBeta_();
+      }
+      if (event.gamma !== this.gammaValue_) {
+        this.gammaValue_ = event.gamma;
+        this.triggerGamma_();
+      }
+    }
+  }
+
+  /**
+   * Dispatches the `alpha` event to signify change in the device orientation
+   * along alpha axis.
+   * @private
+   */
+  triggerAlpha_() {
+    const name = 'alpha';
```

Comment:
If these functions perform identical computations, they can be combined into one function that accepts the name of the value, the value, and the range as parameters, instead of hardcoding the identifiers in each.